### PR TITLE
Fixes #32403 - Add validations to a setting DSL

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Add ability to define validation per Setting through DSL.
It tries to keep similar syntax to rails model validations.

~~Blocked by the DSL definition PR #8437~~

### CODE REVIEW

**Comment by lzap** [comment]
Looks good, have the DSL covered all the required features? In that case can you actually convert all settings to the new DSL? Maybe for starters just one small caregory to iron out issues and once this is RTM you can add the rest to avoid conflicts.

I think we should consider also adding types and maybe few more flags, but that could be done in a separate PR for sure, not here.

**Comment by ezr-ondrej** [comment]
[test upgrade]

**Comment by ezr-ondrej** [comment]
[test foreman]

File: app/models/setting.rb [hunk 1]
```diff
@@ -30,16 +30,20 @@ def validate(record)
     end
   end
 
+  def self.blank_attrs
```
Comment (ekohl) [inline app/models/setting.rb]: Can't we make this something like `blank_attr?(name)` because I feel like that would be more efficient and perhaps easier to read as well.

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 2]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
 The basic types supported by Foreman are: `:boolean`, `:integer`, `:float`, `:string`, `:text`, `:hash`, `:array`.
 The `:text` type supports markdown and usage of such setting should expect markdown syntax when using it.
 
+==== Validate setting value
+
+To validate setting value, you can use API, that tries to mimic the API of ActiveRecord.
+We can use most of the perks offered by ActiveRecord, only defining on setting name instead of attribute.
+We are adding just some shothands like direct regexp validations.
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]:
```suggestion
We are adding just some shorthands like direct regexp validations.
```

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 3]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
 The basic types supported by Foreman are: `:boolean`, `:integer`, `:float`, `:string`, `:text`, `:hash`, `:array`.
 The `:text` type supports markdown and usage of such setting should expect markdown syntax when using it.
 
+==== Validate setting value
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]: This doesn't read that well. I think something like this is better English
```suggestion
==== Validate a setting's value
```

Or
```suggestion
==== Validate setting values
```

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 4]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
...
+
+To validate setting value, you can use API, that tries to mimic the API of ActiveRecord.
+We can use most of the perks offered by ActiveRecord, only defining on setting name instead of attribute.
+We are adding just some shothands like direct regexp validations.
+The attribute is always `value`, you can't validate anything else as it is the only user input.
+Custom validation methods allow you to run validations against other attributes, but please dont! :)
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]: There are legitimate use cases for this. For example, for console encryption we have a field `..._key` and `..._cert` but if the whole feature is off, it's allowed to be empty. If it's on, it must be set. Right now we have no way to properly express this.

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 5]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
...
+
+You have two ways of define the validations:
+* inline with setting definition
+* using `validates` of `validates_with` methods
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]: Can you make `validates` and `validates_with` methods links? I think they're a Rails concept, but as a new developer I may not know that. Totally not speaking for myself right now ;)

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 6]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
...
+
+There is special option `allow_blank`, that needs to be defined on setting definition.
+This also allow blank default value.
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]:
```suggestion
This also allows a blank default value.
```

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 7]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
...
+....
+settings do
+  category(:cfgmgmt) do
+    # Following definitions are missing defaults and full names for simplification
+
+    setting(:blank_setting, type: :string, description: 'Unnecessary setting', allow_blank: true)
+
+    setting(:cool_setting, type: :string, description: 'Setting with only cool values', validate: /^cool.*/, allow_blank: true)
+
+    setting(:cooler_puppet, type: :integer, description: 'Use Puppet that goes to 11', validate: ->(s) { s.value >= 11 })
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]: Shouldn't this be up to and including 11, but not higher?
```suggestion
    setting(:cooler_puppet, type: :integer, description: 'Use Puppet that goes to 11', validate: ->(s) { s.value <= 11 })
```

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 5 above)
Comment (ezr-ondrej) [reply developer_docs/how_to_create_a_plugin.asciidoc]: I'd rather not as they are mimicing the Rails methods with naming, but instead of `validates :attribute` we have `validates :setting_name` I'm not sure if links would not be missleading, though I can mention the specifics :thinking:

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 7 above)
Comment (ezr-ondrej) [reply developer_docs/how_to_create_a_plugin.asciidoc]: That depends, I was trying to suggest you can be even cooler then 11 :thinking:

File: app/models/setting.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/models/setting.rb]: Indeed ! Makes sense, great suggestion!

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 4 above)
Comment (ezr-ondrej) [reply developer_docs/how_to_create_a_plugin.asciidoc]: Other attributes mean description, full_name, options etc.
I'll make it more clear! :)

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 5 above)
Comment (ekohl) [reply developer_docs/how_to_create_a_plugin.asciidoc]: Explaining that difference is actually a good thing IMHO.

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 7 above)
Comment (ekohl) [reply developer_docs/how_to_create_a_plugin.asciidoc]: The original reference of "goes to 11" is an amplifier that has a volume that goes to 11. So it's louder than all the other that go to 10. Of course it's all a joke, but there's a serious note to this: if you understand the joke, you can reason that `validate` will raise an error if the condition evaluates to true instead of false.

File: test/unit/setting_manager_test.rb [hunk 8]
```diff
@@ -83,4 +83,48 @@ class SettingManagerTest < ActiveSupport::TestCase
...
   end
+
+  describe 'Validations' do
+    setup do
+      @validation_backup = Setting._validators[:value].dup
+    end
+
+    teardown do
+      Setting._validators[:value] = @validation_backup
+    end
```
Comment (lzap) [inline test/unit/setting_manager_test.rb:94]: Settings in our tests is in terrible shape - we have fixtures plus stubs. Would be great to have some definitive solution, for example if we would extend minitest/rspec so every test settings get automatically reloaded. Out of scope for this PR tho.

File: test/unit/setting_manager_test.rb [hunk 9]
```diff
@@ -83,4 +83,48 @@ class SettingManagerTest < ActiveSupport::TestCase
...
+            full_name: 'Foo setting')
+          validates(:validfoo, email: true)
+        end
+      end
+      Foreman.settings.load_definitions
+      setting = Setting.new(name: 'validfoo', default: '-omited-', description: 'Omited', value: 'notanemail')
+      assert_not setting.valid?
+    end
+
+    # If it really works, should get tested for specific settings
```
Comment (lzap) [inline test/unit/setting_manager_test.rb]: Please add a postitive test for each type here.

File: test/unit/setting_manager_test.rb (hunk 8 above)
Comment (ezr-ondrej) [reply test/unit/setting_manager_test.rb:94]: All of the DSL defined settings gets loaded and it's default value will be automatically used. I believe that finishing this, migrate all settings to use the DSL and dropping the fixtures all together is the best path forward here :)

File: app/models/setting.rb [hunk 10]
```diff
@@ -30,16 +30,21 @@ def validate(record)
...
+    BLANK_ATTRS.include?(name) || (Foreman::SettingManager.settings.key?(name) && Foreman::SettingManager.settings[name][:options][:allow_blank])
+  end
+
   validates_lengths_from_database
 
   validates :name, :presence => true, :uniqueness => true
   validates :description, :presence => true
-  validates :default, :presence => true, :unless => proc { |s| s.settings_type == "boolean" || BLANK_ATTRS.include?(s.name) }
+  validates :default, :presence => true, :unless => proc { |s| s.settings_type == "boolean" || s.class.allows_blank?(s.name) }
   validates :default, :inclusion => {:in => [true, false]}, :if => proc { |s| s.settings_type == "boolean" }
```
Comment (tbrisker) [inline app/models/setting.rb]: Not directly related, but should we be validating the default values at all? going forward there won't be a default in the db so the validation will never trigger.

File: app/registries/foreman/setting_manager.rb [hunk 11]
```diff
@@ -71,6 +79,37 @@ def setting(name, default:, description:, type:, full_name: nil, value: nil, col
...
+        args << options
+        Setting.validates_with(*args, &block)
+      end
+
+      def _wrap_validation_if(setting_name, options)
+        options[:if] = [
+          ->(setting) { setting.name == setting_name.to_s },
+          *options[:if],
+        ]
+        options[:allow_blank] = storage[setting_name.to_s].dig(:options, :allow_blank) || false
```
Comment (tbrisker) [inline app/registries/foreman/setting_manager.rb]: This line doesn't seem related to this function

**Review by tbrisker — CHANGES_REQUESTED** [review]
Looks good in general, a couple of minor comments inline, and @lzap's request regarding tests hasn't been addressed. 
The code itself is quite complex as we try to cover multiple different ways of defining validations in the DSL, should we support all of them? right now you can also create `validate_setting_name` methods in the setting model that allow more complex custom validations, maybe we could keep those for the complex cases and support only more common cases in the DSL?

File: app/models/setting.rb (hunk 10 above)
Comment (ezr-ondrej) [reply app/models/setting.rb]: I just didn't want to break current behaviour, but I'd be in favor of removing the validations for `default`, as it only validates code, not user input.

**Comment by ezr-ondrej** [comment]
We can leave out a regex validation, but IMHO we need the proc and we should enable the simple validations.
The validations on the class would need to be added from some initializer directly on the `Setting` class, which seems rather ugly to me, the proc validation seems quite elegant to me.

File: app/models/setting.rb (hunk 10 above)
Comment (ekohl) [reply app/models/setting.rb]: I actually like it as a safeguard for developers. However, that should probably be done when loading the settings rather than as a DB level change. So probably in `SettingManager` instead.

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 12]
```diff
@@ -1678,6 +1678,42 @@ The settings are strongly typed and you have to define it.
...
+  category(:cfgmgmt) do
+    # Following definitions are missing defaults and full names for simplification
+
+    setting(:blank_setting, type: :string, description: 'Unnecessary setting', allow_blank: true)
+
+    setting(:cool_setting, type: :string, description: 'Setting with only cool values', validate: /^cool.*/, allow_blank: true)
+
+    setting(:cooler_puppet, type: :integer, description: 'Use Puppet that goes to 11', validate: ->(s) { s.value <= 11 })
+
+    validates(:cooler_puppet, numericality: { greater_than: 11 }, allow_blank: true, if: -> { Setting[:cool_setting] == 'ultra_cool' })
```
Comment (ekohl) [inline developer_docs/how_to_create_a_plugin.asciidoc]: This example is funny. It conflicts with the previous validations. `cool_setting` should never be a valid value. Then `> 11` also conflicts with `<= 11`.

File: app/models/setting.rb (hunk 10 above)
Comment (ezr-ondrej) [reply app/models/setting.rb]: It is there.. you need to pass a default :)
There is no type validation on it though

File: app/registries/foreman/setting_manager.rb (hunk 11 above)
Comment (ezr-ondrej) [reply app/registries/foreman/setting_manager.rb]: I've refactored it

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 12 above)
Comment (ezr-ondrej) [reply developer_docs/how_to_create_a_plugin.asciidoc]: indeed :) changed to `>10` so only possible value is `11` condition is `cool_setting == 'coolest'` so now it should be achievable state :)

File: app/registries/foreman/setting_manager.rb [hunk 13]
```diff
@@ -71,6 +79,42 @@ def setting(name, default:, description:, type:, full_name: nil, value: nil, col
...
+        options = args.extract_options!
+        _add_default_validation_options(name, options)
+        options[:attributes] = [:value]
+        args << options
+        Setting.validates_with(*args, &block)
+      end
+
+      def _add_default_validation_options(setting_name, options)
+        _wrap_validation_if(setting_name, options)
+        options[:allow_blank] = true
```
Comment (tbrisker) [inline app/registries/foreman/setting_manager.rb]: Following our discussion, let's drop this completely and let the user specify it if needed by using `validates` or `validates_with`

File: app/models/setting.rb [hunk 14]
```diff
@@ -34,12 +34,9 @@ def validate(record)
 
   validates :name, :presence => true, :uniqueness => true
   validates :description, :presence => true
-  validates :default, :presence => true, :unless => proc { |s| s.settings_type == "boolean" || BLANK_ATTRS.include?(s.name) }
-  validates :default, :inclusion => {:in => [true, false]}, :if => proc { |s| s.settings_type == "boolean" }
   validates :value, :numericality => true, :length => {:maximum => 8}, :if => proc { |s| s.settings_type == "integer" }
   validates :value, :numericality => {:greater_than => 0}, :if => proc { |s| NONZERO_ATTRS.include?(s.name) }
   validates :value, :inclusion => {:in => [true, false]}, :if => proc { |s| s.settings_type == "boolean" }
-  validates :value, :presence => true, :if => proc { |s| s.settings_type == "array" && !BLANK_ATTRS.include?(s.name) }
```
Comment (tbrisker) [inline app/models/setting.rb:42]: Please add a comment above the constant that it is deprecated and validating that arrays are not empty should be done using the DSL (maybe also clear everything from it?)

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 15]
```diff
@@ -1678,6 +1678,38 @@ The settings are strongly typed and you have to define it.
...
+
+You have two ways to define the validations:
+* inline with setting definition by symbol matching ActiveRecord validator, RegExp on strings, or lambda function that gets value to validate as argument.
+* using `validates` of `validates_with` methods, that mimic [Rails validation methods](https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html), but are using setting names instead of attribute names, as the validated attribute is always value in this case.
+
+[source, ruby]
+....
+settings do
+  category(:cfgmgmt) do
+    # Following definitions are missing defaults and full names for simplification
```
Comment (tbrisker) [inline developer_docs/how_to_create_a_plugin.asciidoc]:
```suggestion
    # Following definitions are missing full names for simplification
```

File: developer_docs/how_to_create_a_plugin.asciidoc [hunk 16]
```diff
@@ -1678,6 +1678,38 @@ The settings are strongly typed and you have to define it.
...
+
+    setting(:blank_setting, type: :string,  default: '', description: 'Unnecessary setting')
+
+    setting(:cool_setting, type: :string, default: 'cool' description: 'Setting with only cool values', validate: /^cool.*/)
+
+    setting(:cooler_puppet, type: :integer, default: 5, description: 'Use Puppet that goes to 11', validate: ->(value) { value <= 11 })
+
+    validates(:cooler_puppet, numericality: { greater_than: 10 }, if: -> { Setting[:cool_setting] == 'coolest' })
```
Comment (tbrisker) [inline developer_docs/how_to_create_a_plugin.asciidoc]: maybe
```suggestion
    validates(:cooler_puppet, numericality: { greater_than: 10 }, if: -> { Setting[:cool_setting] == 'coolest' }, allow_blank: true)
```
to show how it is possible to pass allow_blank?

File: app/registries/foreman/setting_manager.rb [hunk 17]
```diff
@@ -71,6 +79,42 @@ def setting(name, default:, description:, type:, full_name: nil, value: nil, col
...
+
+      def validates(name, validations)
+        if validations.is_a?(Proc)
+          validates_with name, ValueLambdaValidator, proc: validations
+          return
+        elsif validations.is_a?(Regexp)
+          validations = { format: { with: validations } }
+        elsif validations.is_a?(Symbol)
+          validations = { validations => true }
+        end
```
Comment (tbrisker) [inline app/registries/foreman/setting_manager.rb:94]: if we want this method to mimic the activerecord API, this part probably belongs outside it?

**Review by tbrisker — DISMISSED** [review]

File: app/registries/foreman/setting_manager.rb (hunk 17 above)
Comment (ezr-ondrej) [reply app/registries/foreman/setting_manager.rb:94]: well I'd like to allow `setting('name', validate: :email)` and thats only shorthand for `setting('name'); validates('name', :email)`. But I've moved it to enable this only in the inline validations.

File: developer_docs/how_to_create_a_plugin.asciidoc (hunk 16 above)
Comment (ezr-ondrej) [reply developer_docs/how_to_create_a_plugin.asciidoc]: it doesn't make much sense here, but yeah, sure :)

**Comment by ezr-ondrej** [comment]
[test foreman] what the heck...

**Comment by ezr-ondrej** [comment]
[test foreman][test katello][test upgrade]

**Comment by ezr-ondrej** [comment]
[test foreman]

**Comment by tbrisker** [comment]
@ezr-ondrej looks like a test failure related to the change, i think the test needs to be updated.

**Comment by ezr-ondrej** [comment]
> @ezr-ondrej looks like a test failure related to the change, i think the test needs to be updated.

Well they were not able to run prior this, there's issue in connection to GH from Jenkins. But yeah, this is related.

**Comment by lzap** [comment]
I am unassigning myself from this, Tomer has a lot more to say. Generally, I don't feel comfortable with reviewing more advanced RoR (ActiveRecord) stuff.

**Comment by ezr-ondrej** [comment]
> @ezr-ondrej looks like a test failure related to the change, i think the test needs to be updated.

I've just removed the test as it is pointless now :)

**Review by ares — APPROVED** [review]
Thanks @ezr-ondrej, looks good to me, I also tested locally and it seems to work fine. I think this is demo-worthy for devs. Thanks @tbrisker for doing the main review, I'm merging now.

**Comment by ares** [comment]
[test katello] just to be sure since I didn't see these 3 failure elsewhere

**Comment by ezr-ondrej** [comment]
:green_apple:

**Comment by ares** [comment]
Thansk again!
